crates/feldera-storage/src/backend/mod.rs

```diff
@@ -0,0 +1,244 @@
+#[derive(Error, Debug)]
+pub enum StorageError {
+    #[cfg(feature = "glommio")]
+    #[error("Got IO error during glommio operation")]
+    Io(#[from] glommio::GlommioError<()>),
+    #[error("Got IO error during monoio operation")]
+    StdIo(#[from] std::io::Error),
+}
+
+/// A trait for a storage backend to implement so client can create/delete
+/// files.
+pub trait StorageControl {
+    /// Creates a new persistent file used for writing data.
+    ///
+    /// Returns a file-descriptor that can be used for writing data.
+    /// Note that it is not possible to read from this file until
+    /// [`StorageWrite::complete`] is called and the [`FileHandle`] is
+    /// converted to an [`ImmutableFileHandle`].
+    async fn create(&self) -> Result<FileHandle, StorageError>;
+
+    /// Deletes a previously completed file.
+    ///
+    /// This removes the file from the storage backend and makes it unavailable
+    /// for reading.
+    async fn delete(&self, fd: ImmutableFileHandle) -> Result<(), StorageError>;
+
+    /// Deletes a previously created file.
+    ///
+    /// This removes the file from the storage backend and makes it unavailable
+    /// for writing.
+    ///
+    /// Use [`delete`] for deleting a file that has been completed.
+    async fn delete_mut(&self, fd: FileHandle) -> Result<(), StorageError>;
+}
+
+/// A trait for a storage backend to implement so clients can write to files.
+pub trait StorageWrite {
+    /// Allocates a buffer suitable for writing to a file using Direct I/O over
+    /// `io_uring`.
+    fn allocate_buffer(sz: usize) -> WriteBuffer;
+
+    /// Writes a block of data to a file.
+    ///
+    /// ## Arguments
+    /// - `fd` is the file-handle to write to.
+    /// - `offset` is the offset in the file to write to.
+    /// - `data` is the data to write.
+    ///
+    /// ## Preconditions
+    /// - `data.len() >= 512`: Minimum block size for Direct I/O, otherwise the
+    ///   write will return an error.
+    /// - `data.len().is_power_of_two() == true`: The size of the buffer must be
+    ///   a power of two.
+    /// - `offset.is_power_of_two() == true`: The offset must be a power of two.
+    ///
+    /// ## Returns
+    /// The number of bytes written.
+    ///
+    /// API returns an error if any of the above preconditions are not met.
+    async fn write_block(
+        &self,
+        fd: &FileHandle,
+        offset: u64,
+        data: WriteBuffer,
+    ) -> Result<usize, StorageError>;
+
+    /// Completes writing of a file.
+    ///
+    /// This makes the file available for reading by returning a file-descriptor
+    /// that can be used for reading data.
+    ///
+    /// ## Arguments
+    /// - `fd` is the file-handle to complete.
+    ///
+    /// ## Returns
+    /// A file-descriptor that can be used for reading data. See also
+    /// [`StorageRead`].
+    async fn complete(&self, fd: FileHandle) -> Result<ImmutableFileHandle, StorageError>;
```

Comment:
I'm surprised this doesn't consume the `StorageWrite`.

Reply:
It currently consumes the `FileHandle`, and transforms it into a ImmutableFileHandle (FileHandle (and ImmutableFileHandle aren't Clone) so you can be sure you lost the ability to write to the fd after `complete()`)

But we can discuss a way to consume the StorageWrite too if you think that's better

Reply:
I think I see. This seems fine.